[CELEBORN-1174] Introduce application dimension resource consumption metrics

SteNicholas opened the pull request:
### What changes were proposed in this pull request?

Introduce application dimension resource consumption metrics for `ResourceConsumptionSource`.

### Why are the changes needed?

`ResourceConsumption` namespace metrics are generated for each user and they are identified using a metric tag at present. It's recommended to introduce application dimension resource consumption metrics that expose application dimension resource consumption of Master and Worker. By monitoring resource consumption in the application dimension, you can obtain the actual situation of application resource consumption. 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

- `WorkerInfoSuite#WorkerInfo toString output`
- `PbSerDeUtilsTest#fromAndToPbResourceConsumption`
- `MasterStateMachineSuitej#testObjSerde`

RexXiong commented:
Since Celeborn supports multi-tenancy, I suggest that the app's resource consumption should also be categorized by tenant-sub-user.

FMX commented on worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:
```diff
@@ -565,14 +614,30 @@ private[celeborn] class Worker(
     resourceConsumption
   }
 
+  private def computeAppResourceConsumption(appResourceConsumption: (
```
IMO, these method looks quite similar. I think they can be merged.

FMX commented on worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:
```diff
@@ -528,35 +546,66 @@ private[celeborn] class Worker(
   }
 
-  private def handleResourceConsumption(): util.Map[UserIdentifier, ResourceConsumption] = {
+  private def handleUserResourceConsumption(): util.Map[UserIdentifier, ResourceConsumption] = {
```
This method handleUserResourceConsumption is similar to handleAppResourceConsumption. They can be merged.

FMX commented on worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StorageManager.scala:
```diff
@@ -839,6 +839,38 @@ final private[worker] class StorageManager(conf: CelebornConf, workerSource: Abs
     }
   }
 
+  def appResourceConsumptionSnapshot(): Map[String, ResourceConsumption] = {
```
This method is similar to userResourceConsumptionSnapshot.

FMX commented on common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:
```diff
@@ -188,6 +194,14 @@ class WorkerInfo(
     userResourceConsumption
   }
 
+  def updateThenGetAppResourceConsumption(consumption: util.Map[
```
I think we can reduce duplicate codes snippets.

FMX reviewed (commented):
Will the grafana dashboard be updated according to this PR?

SteNicholas commented:
@FMX, I have addressed above comments and updated `celeborn-dashboard.json` for application dimension resource consumption metrics.  PTAL.
@onebox-li, could you help to review whether the dashboard of `celeborn-dashboard.json` is correct?

SteNicholas commented:
@RexXiong, I have supported that resource consumption of application be categorized by tenant-sub-user via `ApplicationIdentifier` which includes the tenantId and name of the user identifier. PTAL.
@FMX, I have improved duplicated codes snippets and addressed above comment. PTAL.
@onebox-li, could you help to review the celeborn dashboard?

FMX commented on common/src/main/scala/org/apache/celeborn/common/identity/ApplicationIdentifier.scala:
```diff
@@ -0,0 +1,58 @@
+import org.apache.celeborn.common.exception.CelebornException
+import org.apache.celeborn.common.internal.Logging
+
+case class ApplicationIdentifier(tenantId: String, name: String, applicationId: String) {
```
Does this modification match your expectations? @RexXiong

RexXiong replied:
> Does this modification match your expectations? @RexXiong

As we already have UserResourceConsumption，ApplicationIdentifier for APPResourceConsumption seems redundant，I think what we actually need is a nested structure, like map<string, map<string, ResourceConsumption>>, thanks.

RexXiong commented on common/src/main/proto/TransportMessages.proto:
```diff
@@ -162,8 +163,9 @@ message PbRegisterWorker {
-  string requestId = 9;
   map<string, PbResourceConsumption> userResourceConsumption = 8;
+  string requestId = 9;
+  map<string, PbResourceConsumption> applicationResourceConsumption = 10;
```
IMO the key of applicationResourceConsumption does not look so intuitive，
At first look,  I thought this must be the applicationId.

SteNicholas commented:
@RexXiong, @FMX, I have followed your suggestion which uses `subResourceConsumption` to store resource consumpution of  application. PTAL.

RexXiong commented on common/src/main/scala/org/apache/celeborn/common/quota/ResourceConsumption.scala:
```diff
@@ -17,26 +17,67 @@
+    if (subResourceConsumptions == null) {
+      resourceConsumptions = other.subResourceConsumptions
+    } else {
+      resourceConsumptions = subResourceConsumptions
```
If use subResourceConsumptions directly， IMO it may lead to repeated resource consumption calculations.

RexXiong commented on master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:
```diff
@@ -851,55 +851,74 @@ private[celeborn] class Master(
-
+      applicationId: String = null): Unit = {
+    val resourceConsumptionLabel =
+      if (applicationId == null) userIdentifier.toMap
```
IMO we should clean these metrics regularly, such as when the application timeout?

SteNicholas commented:
@RexXiong, I have addressed above comment and added the cleaner of application resource consumption gauges. PTAL.

RexXiong commented on worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:
```diff
@@ -596,6 +625,38 @@ private[celeborn] class Worker(
     }
 
+  private def removeAppResourceConsumption(applicationId: String): Unit = {
+    userResourceConsumptions.asScala.foreach {
```
userResourceConsumptions may not contain the application info ,then The gauge of the application in ResourceConsumptionSource can't be removed

SteNicholas commented:
@RexXiong, I have addressed comments mentioned above and rebased main branch to update the resource consumption gauge of worker. PTAL.

RexXiong reviewed (approved):
LGTM, and user level resource consumption metrics still left, IMO it should be clean too, (this can be done in future..

AngersZhuuuu reviewed (approved):
LGTM

FMX reviewed (approved):
LGTM

SteNicholas commented:
Merged into main(v0.5.0).
